File: archive/link-api/app/models/link.py
from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING, Optional
from uuid import UUID, uuid4

from sqlalchemy import ForeignKey, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base

if TYPE_CHECKING:
    from app.models.user import User


class Link(Base):
    __tablename__ = "links"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    url: Mapped[str]
    title: Mapped[Optional[str]]
    description: Mapped[Optional[str]]
    text: Mapped[Optional[str]]
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now()
    )

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="links")

    # Constraints
    __table_args__ = (UniqueConstraint("user_id", "url", name="unique_user_url"),)

File: link-mgmt-go/pkg/models/link.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Link struct {
	ID          uuid.UUID `db:"id" json:"id"`
	UserID      uuid.UUID `db:"user_id" json:"user_id"`
	URL         string    `db:"url" json:"url"`
	Title       *string   `db:"title" json:"title,omitempty"`
	Description *string   `db:"description" json:"description,omitempty"`
	Text        *string   `db:"text" json:"text,omitempty"`
	CreatedAt   time.Time `db:"created_at" json:"created_at"`
	UpdatedAt   time.Time `db:"updated_at" json:"updated_at"`
}

// LinkCreate represents data for creating a new link
type LinkCreate struct {
	URL         string  `json:"url" binding:"required"`
	Title       *string `json:"title,omitempty"`
	Description *string `json:"description,omitempty"`
	Text        *string `json:"text,omitempty"`
}

// LinkUpdate represents data for updating a link
type LinkUpdate struct {
	URL         *string `json:"url,omitempty"`
	Title       *string `json:"title,omitempty"`
	Description *string `json:"description,omitempty"`
	Text        *string `json:"text,omitempty"`
}
